component/xload/comp/lister.go

```diff
@@ -0,0 +1,198 @@
+	if rl.GetThreadPool() != nil {
+		rl.GetThreadPool().Stop()
+	}
+	rl.GetNext().Stop()
+}
+
+// wait for the configured block-list-on-mount-sec to make the list call
+func waitForListTimeout() error {
+	var blockListSeconds uint16 = 0
+	err := config.UnmarshalKey("azstorage.block-list-on-mount-sec", &blockListSeconds)
+	if err != nil {
+		return err
+	}
+	time.Sleep(time.Duration(blockListSeconds) * time.Second)
+	return nil
+}
+
+func (rl *remoteLister) Process(item *common.WorkItem) (int, error) {
+	absPath := item.Path // TODO:: xload : check this for subdirectory mounting
+
+	log.Debug("remoteLister::Process : Reading remote dir %s", absPath)
+
+	// this block will be executed only in the first list call for the remote directory
+	// so haven't made the listBlocked variable atomic
+	if !rl.listBlocked {
+		log.Debug("remoteLister::Process : Waiting for block-list-on-mount-sec before making the list call")
+		err := waitForListTimeout()
+		if err != nil {
+			log.Err("remoteLister::Process : unable to unmarshal block-list-on-mount-sec [%s]", err.Error())
+			return 0, err
+		}
+		rl.listBlocked = true
+	}
+
+	marker := ""
+	var cnt, iteration int
+	for {
+		entries, new_marker, err := rl.GetRemote().StreamDir(internal.StreamDirOptions{
+			Name:  absPath,
+			Token: marker,
+		})
+		if err != nil {
+			log.Err("remoteLister::Process : Remote listing failed for %s [%s]", absPath, err.Error())
+		}
+
+		marker = new_marker
+		cnt += len(entries)
+		iteration++
+		log.Debug("remoteLister::Process : count: %d , iterations: %d", cnt, iteration)
+
+		for _, entry := range entries {
+			log.Debug("remoteLister::Process : Iterating: %s, Is directory: %v", entry.Path, entry.IsDir())
+
+			if entry.IsDir() {
+				// create directory in local
+				// spawn go routine for directory creation and then
+				// adding to the input channel of the listing component
+				// TODO:: xload : check how many threads can we spawn
+				go func(name string) {
+					localPath := filepath.Join(rl.path, name)
+					err = rl.mkdir(localPath)
+					// TODO:: xload : handle error
+					if err != nil {
+						log.Err("remoteLister::Process : Failed to create directory [%s]", err.Error())
+						return
+					}
+
+					// push the directory to input pool for its listing
+					rl.GetThreadPool().Schedule(&common.WorkItem{
+						CompName: rl.GetName(),
+						Path:     name,
+					})
+				}(entry.Path)
+			} else {
+				// send file to the output channel for chunking
+				rl.GetNext().GetThreadPool().Schedule(&common.WorkItem{
+					CompName: rl.GetNext().GetName(),
+					Path:     entry.Path,
+					DataLen:  uint64(entry.Size),
+				})
+			}
+		}
+
+		if len(new_marker) == 0 {
+			log.Debug("remoteLister::Process : remote listing done for %s", absPath)
+			break
+		}
+	}
+
+	return cnt, nil
+}
+
+func (rl *remoteLister) mkdir(name string) error {
+	log.Debug("remoteLister::mkdir : Creating local path: %s", name)
+	return os.MkdirAll(name, 0777)
```

Comment:
If we are doing only mkdir in this function than better to remove the function call overhead and do it in the process method only.

Reply:
Just trying to separate the flow functionality wise.

Reply:
Also having this as a separate function, so that when the support for upload is added, this method will have logic to create directory in the storage.